azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunner.java

```diff
@@ -1091,20 +1098,88 @@ public int getExecutionId() {
     return this.execId;
   }
 
+  public Set<JobRunner> getActiveJobRunners() {
+    return ImmutableSet.copyOf(this.activeJobRunners);
+  }
+
+  // Class helps report the flow start and stop events.
+  private class FlowRunnerEventListener implements EventListener {
+
+    public FlowRunnerEventListener() {
+    }
+
+    private synchronized Map<String, String> getFlowMetadata(final FlowRunner flowRunner) {
+      final ExecutableFlow flow = flowRunner.getExecutableFlow();
+      final Map<String, String> metaData = new HashMap<>();
+      metaData.put("flowName", flow.getId());
+      metaData.put("azkabanHost",
+          flowRunner.azkabanProps.getString("azkaban.server.hostname", "unknown"));
+      metaData.put("projectName", flow.getProjectName());
+      metaData.put("submitUser", flow.getSubmitUser());
+      metaData.put("executionId", String.valueOf(flow.getExecutionId()));
+      metaData.put("startTime", String.valueOf(flow.getStartTime()));
+      metaData.put("submitTime", String.valueOf(flow.getSubmitTime()));
+      return metaData;
+    }
+
+    @Override
+    public synchronized void handleEvent(final Event event) {
+      if (FlowRunner.this.azkabanEventReporter == null) {
+        FlowRunner.this.logger.debug("Event reporter is not enabled");
+        return;
+      }
+
+      if (event.getType() == Type.FLOW_STARTED) {
+        final FlowRunner flowRunner = (FlowRunner) event.getRunner();
+        final ExecutableFlow flow = flowRunner.getExecutableFlow();
+        FlowRunner.this.logger.info("Flow started: " + flow.getId());
+        FlowRunner.this.azkabanEventReporter.report(event, getFlowMetadata(flowRunner));
+      } else if (event.getType() == Type.FLOW_FINISHED) {
+        final FlowRunner flowRunner = (FlowRunner) event.getRunner();
+        final ExecutableFlow flow = flowRunner.getExecutableFlow();
+        FlowRunner.this.logger.info("Flow ended: " + flow.getId());
+        final Map<String, String> flowMetadata = getFlowMetadata(flowRunner);
+        flowMetadata.put("endTime", String.valueOf(flow.getEndTime()));
+        FlowRunner.this.azkabanEventReporter.report(event, flowMetadata);
+      }
+    }
+  }
+
   private class JobRunnerEventListener implements EventListener {
 
     public JobRunnerEventListener() {
     }
 
+    private synchronized Map<String, String> getJobMetadata(final JobRunner jobRunner) {
+      final ExecutableNode node = jobRunner.getNode();
+      final Map<String, String> metaData = new HashMap<>();
+      metaData.put("jobId", node.getId());
```

Comment:
The keys in the metaData become part of the contract. Shall they be defined in an interface?

Reply:
Metadata can be treated as a blackbox by the reporter.

Reply:
Please define how the metadata is supposed to be used. 
The schema of the metadata is still part of the contract between the producer (azkaban) and the consumers of this information. Right? If so, where is this contract defined?

Reply:
Are you suggesting to use an interface with constants? This is generally considered a bad practice. Refer [this](https://stackoverflow.com/questions/320588/interfaces-with-static-fields-in-java-for-sharing-constants/320642#320642). As part of the metadata consumers will get a map which is what we mentioned in the interface AzkabanEventReporter.

Reply:
Yes, I've read the Effective Java book and remember this recommendation. My understanding is that it is not a problem if the constants are part of the contract.